### PR TITLE
BUG: invalid cm/tm in visitor functions

### LINKED ISSUE #2059: BUG: Incorrect text matrix passed to visitor_text in page.extract_text
When supplying `page.extract_text` with a visitor_text function, the callback to the function is made with an incorrect `tm_matrix` (text matrix) parameter.

I believe this is due to the fact that [here](https://github.com/py-pdf/pypdf/blob/c04a6bba997d353dbe4efe160596472d36c07604/pypdf/_text_extraction/__init__.py#L117), the visitor is called with the new `tm_matrix`, even though at this point, the text matrix may already be changed by the currently handled OP.

## Environment

Running the current `main` branch of pypdf on Debian.

```bash
$ python -m platform
Linux-5.10.0-23-amd64-x86_64-with-glibc2.31

$ python -c "import pypdf;print(pypdf.__version__)"
3.14.0
```

## Code + PDF

Here is an example script that converts a [testdoc.pdf](https://github.com/py-pdf/pypdf/files/12249574/testdoc.pdf) to SVG, logs the OP's in the page and the calls to the visitor with x, y coordinates.

```python
from pypdf import PdfReader
import svgwrite

file_name = "./testdoc.pdf"

reader = PdfReader(file_name)
page = reader.pages[0]

print("####\nOps:\n####")
print(page.get_contents().operations)

length, height = page.mediabox[2:]

dwg = svgwrite.Drawing(file_name[:-3] + "svg", size=(length, height), profile="full")


def visitor_svg_text(text, cm, tm, fontDict, fontSize):
    (x, y) = (tm[4], tm[5])
    print(x, y, text)
    dwg.add(dwg.text(text, insert=(x, height-y), fill="blue", style=f"font-size: {fontSize}px"))


print("\n####\nParsed Lines:\n####")
page.extract_text(visitor_text=visitor_svg_text)
dwg.save()
```

Output:
```txt
####
Ops:
####
[([], b'BT'), (['/F29', 14.3462], b'Tf'), ([133.768, 707.125], b'Td'), ([['1', -1125, 'T', 94, 'estsection']], b'TJ'), (['/F19', 9.9626000000000001], b'Tf'), ([0, -21.821000000000002], b'Td'), ([['A', -333, 'B', -334, 'C']], b'TJ'), ([169.36500000000001, -546.04899999999998], b'Td'), ([['1']], b'TJ'), ([], b'ET')]

####
Parsed Lines:
####
0.0 0.0 
133.768 707.125 1 Testsection
133.768 685.304 

303.13300000000004 139.255 A B C

303.13300000000004 139.255 1
```

As can be seen (also in the resulting SVG), the coordinates of the line "A B C" get affected by the `Td` following *after it*, because this is when  `crlf_space_check` detects a new line. It then however supplies the `tm_matrix` that was already altered to the call to the visitor, instead of the one that was active before the operator was applied.

### LINKED ISSUE #2075: x values in the tm_matrix are wrong
I am trying to read what seems to be a not very complex pdf. Here is a bit from one page:

Screenshot from 2023-08-09 17-44-08

I am pulling out the y and then x value from the tm_matrix and the text from the visitor_text. I am getting this:

```
[...]
{'text_matrix': [1, 0, 0, 1, 26, 541], 'text': 'ADOPTION \n'}
{'text_matrix': [1, 0, 0, 1, 123, 530], 'text': 'adults, adoption of,  AB 1756 '}
{'text_matrix': [1, 0, 0, 1, 273, 519], 'text': 'agencies, organizations, etc.: requirements, prohibitions, etc.,  SB 807 '}
{'text_matrix': [1, 0, 0, 1, 245, 508], 'text': 'assistance programs, adoption: nonminor dependents,  SB 9 '}
{'text_matrix': [1, 0, 0, 1, 114, 497], 'text': 'birth certificates,  AB 1302 '}
{'text_matrix': [1, 0, 0, 1, 35, 486], 'text': 'contact agreements, postadoption— \n'}
{'text_matrix': [1, 0, 0, 1, 110, 474], 'text': 'birth parents,  AB 1650 '}
{'text_matrix': [1, 0, 0, 1, 93, 463], 'text': 'siblings,  AB 20 '}
{'text_matrix': [1, 0, 0, 1, 130, 452], 'text': 'facilitators, adoption,  AB 120'}
{'text_matrix': [1, 0, 0, 1, 164, 452], 'text': ';  SB 120'}
{'text_matrix': [1, 0, 0, 1, 184, 452], 'text': ',  807 '}
{'text_matrix': [1, 0, 0, 1, 199, 441], 'text': 'failed adoptions: reproductive loss leave,  SB 848 '}
{'text_matrix': [1, 0, 0, 1, 300, 430], 'text': 'hearings, adoption finalization: remote proceedings, technology, etc.,  SB 21 '}
{'text_matrix': [1, 0, 0, 1, 135, 419], 'text': 'native american tribes,  AB 120'}
{'text_matrix': [1, 0, 0, 1, 168, 419], 'text': ';  SB 120 '}
{'text_matrix': [1, 0, 0, 1, 170, 408], 'text': 'parental rights, reinstatement of,  AB 20 '}
{'text_matrix': [1, 0, 0, 1, 265, 397], 'text': 'parents, prospective adoptive: criminal background checks,  SB 824 '}
{'text_matrix': [1, 0, 0, 1, 26, 386], 'text': 'ADULT EDUCATION \n'}
{'text_matrix': [1, 0, 0, 1, 150, 375], 'text': 'services, adult educational,  SB 877 '}
{'text_matrix': [1, 0, 0, 1, 140, 364], 'text': 'week, adult education,  ACR 31 '}
{'text_matrix': [1, 0, 0, 1, 26, 353], 'text': 'ADVERTISING. See also MARKETING; and particular subject matter (e.g., \n'}
{'text_matrix': [1, 0, 0, 1, 68, 342], 'text': 'ELECTIONS). \n'}
{'text_matrix': [1, 0, 0, 1, 211, 331], 'text': 'alcoholic beverages: tied-house restrictions,  AB 546'}
{'text_matrix': [1, 0, 0, 1, 231, 331], 'text': ',  840'}
{'text_matrix': [1, 0, 0, 1, 251, 331], 'text': ',  1294'}
{'text_matrix': [1, 0, 0, 1, 290, 331], 'text': ' ;  SB 392'}
{'text_matrix': [1, 0, 0, 1, 310, 331], 'text': ',  430 '}
{'text_matrix': [1, 0, 0, 1, 206, 320], 'text': 'campaign re social equity, civil rights, etc.,  SB 447 '}
{'text_matrix': [1, 0, 0, 1, 87, 309], 'text': 'cannabis,  AB 794'}
{'text_matrix': [1, 0, 0, 1, 107, 309], 'text': ',  1207 '}
{'text_matrix': [1, 0, 0, 1, 35, 298], 'text': 'elections. See ELECTIONS. \n'}
{'text_matrix': [1, 0, 0, 1, 35, 287], 'text': 'false, misleading, etc., advertising— \n'}
{'text_matrix': [1, 0, 0, 1, 155, 276], 'text': 'disgorgement, remedy of,  AB 1366 '}
{'text_matrix': [1, 0, 0, 1, 218, 265], 'text': 'master of divinity: prohibited title displays,  AB 1564 '}
{'text_matrix': [1, 0, 0, 1, 232, 254], 'text': 'pregnancy-related services: civil penalties, etc.,  AB 315'}
{'text_matrix': [1, 0, 0, 1, 253, 254], 'text': ',  602 '}
{'text_matrix': [1, 0, 0, 1, 172, 243], 'text': 'pricing for goods and services,  SB 478 '}
{'text_matrix': [1, 0, 0, 1, 321, 232], 'text': 'hotels, short-term rentals, etc., advertised rates: mandatory fee disclosures,  SB 683 '}
{'text_matrix': [1, 0, 0, 1, 247, 221], 'text': 'housing rental properties advertised rates: disclosures,  SB 611 '}
{'text_matrix': [1, 0, 0, 1, 25, 190], 'text': '*2023–24 First Extraordinary Session bills are designated (1X). '}
```

There are only 2 levels of indentation in the text, as you can see from the screenshot. And the x values are all over the place,

The amount of error in the x value seems to be somewhat proportional to the number of spaces that have been lost. I wonder if this is significant.

## Environment

OS: Ubuntu 22.04.2 LTS

     % pip freeze | grep pdf
     pypdf==3.15.0

     $ python -m platform
     Linux-6.2.0-26-generic-x86_64-with-glibc2.35

     $ python -c "import pypdf;print(pypdf.__version__)"
     3.15.0

## Code + PDF?

[LegIndex-page6.pdf](https://github.com/py-pdf/pypdf/files/12318042/LegIndex-page6.pdf)

```python
from pypdf import PdfReader


def text_details(text, curr_trans_matrix, text_matrix, font_dict, font_size):
    info = {
        "text": text,
        "curr_trans_matrix": curr_trans_matrix,
        "text_matrix": text_matrix,
        "font_dict": font_dict,
        "font_size": font_size,
    }

    # put into a dictionary keyed by y value to enable sort.

    if info["text"] != "" and info["text"] != "\u200b" and info["text"] != "\n":
        global strings
        y_val = info["text_matrix"][5]
        if y_val not in strings:
            strings[y_val] = list()
        strings[y_val].append({"text_matrix": [int(el) for el in text_matrix], "text": text})


if __name__ == "__main__":
    path = "LegIndex-page6.pdf"

    strings = {}

    pdf = PdfReader(path)
    text_list = pdf.pages[0].extract_text().split("\n")
    pdf.pages[0].extract_text(visitor_text=text_details)
    y_vals = reversed(sorted(list(strings.keys())))
    for y_val in y_vals:
        for string in strings[y_val]:
            print(string)
```

### LINKED ISSUE #2200: BUG: text position detection works worse than in PyPDF2
I'm trying to add a highlighting annotation to the doc using the text visitor to identify the coordinates to add it.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
macOS-13.5.2-x86_64-i386-64bit

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==3.16.1, crypt_provider=('local_crypt_fallback', '0.0.0'), PIL=10.0.1
```

## Code + PDF

This is a minimal, complete example that shows the issue:

repro repo https://github.com/vors/pypdf-highlighting-repro

Share here the PDF file(s) that cause the issue. The smaller they are, the
better. Let us know if we may add them to our tests!

You can add them to your tests.

## Visuals

image

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2200
closes #2075
reworks and is still valid to close #2059

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@troethe, 
From my analysis, the change you've proposed was not suffcient as the prev position may be overriden during the process.
I've created a PR where I've storing the cm/tm matrices when resetting the text variable (this normally occurs when meeting the Td which will occurs/set the position before restarting a new line
Can you have a review of my change?

**Review by troethe — APPROVED** [review]
Looks good! I think this approach is currently the best way to handle #2075. The two comments I added are minutiae and not necessarily important.

File: pypdf/_text_extraction/__init__.py [hunk 1]
```diff
@@ -100,13 +99,21 @@ def crlf_space_check(
...
+    memo_tm = memo_cmtm[1]
+
     m_prev = mult(tm_prev, cm_prev)
     m = mult(tm_matrix, cm_matrix)
     orientation = orient(m)
     delta_x = m[4] - m_prev[4]
     delta_y = m[5] - m_prev[5]
     k = math.sqrt(abs(m[0] * m[3]) + abs(m[1] * m[2]))
     f = font_size * k
+    cm_prev = m
```
[the remark this answers is not in the document]
Comment (pubpub-zz) [reply pypdf/_text_extraction/__init__.py:116]: agree. 
Good point to have noticed it . Actually I've forgot to pass the multiplied cm_matrix to the visitor

**Comment by pubpub-zz** [comment]
@troethe 
I've taken into account your comments

@srogmann 
I would like to have your opinion too:
I've also completed the PR passing the product of cm_matrix . tm_matrix as the users may not have an easy access to `mult()` function and because cm_matrix alone has no mean (for me)
also for me it is better to recommand to work on the cm_matrix  as it will take into account all transformations()
finally looking at the code I currently see no difference between using `visitor_operand_before()` and `visitor_operand_after()`.
Shouldn't we await from `visitor_operand_before()` to return the tuple operator,operands to allow the user to modify/cancel the processing of some operations ?

**Comment by srogmann** [comment]
@pubpub-zz  I also have no example of cm_matrix without tm_matrix at hand. PDF-reference, "5.3.3 Text Space Details", uses the product of cm_matrix and tm_matrix you mentioned.

An argument to stay at cm_matrix instead of the product might be the performance if the product is computed but not used. In that case the publish of a mult()-function might be useful.

In text-extraction-test 3 of tests/test_page.py I used visitor_operand_after to verify the Td-operation (see visitor_td).

The return of the operator by visit_operand_before() can support patches or studies like the one in https://github.com/py-pdf/pypdf/pull/1389#issuecomment-1281524843. Can or should the operand-list be modified inside visit_operand_before() by using append(...), clear(...), ...? Otherwise it would be better to return the tuple (operator, operands).

**Comment by pubpub-zz** [comment]
> @pubpub-zz I also have no example of cm_matrix without tm_matrix at hand. PDF-reference, "5.3.3 Text Space Details", uses the product of cm_matrix and tm_matrix you mentioned.
> 
> An argument to stay at cm_matrix instead of the product might be the performance if the product is computed but not used. In that case the publish of a mult()-function might be useful.
> 
> In text-extraction-test 3 of tests/test_page.py I used visitor_operand_after to verify the Td-operation (see visitor_td).
> 
> The return of the operator by visit_operand_before() can support patches or studies like the one in [#1389 (comment)](https://github.com/py-pdf/pypdf/pull/1389#issuecomment-1281524843). Can or should the operand-list be modified inside visit_operand_before() by using append(...), clear(...), ...? Otherwise it would be better to return the tuple (operator, operands).

@MartinThoma 
your opinion?

**Comment by MartinThoma** [comment]
I don't have an opinion on this topic. To be honest, I haven't read the specs so I need to trust you on those changes.

If two of you agree that this PR is good (and CI is green), then I would merge :-)

**Comment by pubpub-zz** [comment]
@MartinThoma  / @srogmann 
the PR for me is mature.
I left the change of cm_matrix which now reflects the position/orientation/scale of the text in the page user space : 
this may be considered as a change in the interface but I don't think many people are using the existing matrix and the change makes things easier to handle (the matrix description was very obscure)
I did not implement the new feature to use visit_operand_before to modify the operations. I will do this in an independant PR for traceability.

**Comment by pubpub-zz** [comment]
Plus stdby

**Comment by MartinThoma** [comment]
> Plus stdby

I take this as "please standby", so I should not merge?

**Comment by pubpub-zz** [comment]
> > Plus stdby
> 
> I take this as "please standby", so I should not merge?

Correct interpretation
@jianfan2012 used the visitor in an interesting way where the multiplication of cm.tm is preventing the use
we need to keep then separate. I'm reverting my change update the documentation

**Comment by pubpub-zz** [comment]
@MartinThoma 
All should be good now.

File: docs/user/extract-text.md [hunk 2]
```diff
@@ -27,14 +27,26 @@ Refer to [extract\_text](../modules/PageObject.html#pypdf._page.PageObject.extra
 You can use visitor-functions to control which part of a page you want to process and extract. The visitor-functions you provide will get called for each operator or for each text fragment.
 
 The function provided in argument visitor_text of function extract_text has five arguments:
-text, current transformation matrix, text matrix, font-dictionary and font-size.
-In most cases the x and y coordinates of the current position
-are in index 4 and 5 of the current transformation matrix.
+* text : the current text (as long as possible, can be up to a full line)
+* user_matrix: current matrix to move from user coordinate space(aka. CTM)
```
Comment (MartinThoma) [inline docs/user/extract-text.md:31]:
```suggestion
* user_matrix: current matrix to move from user coordinate space (also known as CTM)
```

File: docs/user/extract-text.md [hunk 3]
```diff
@@ -27,14 +27,26 @@ Refer to [extract\_text](../modules/PageObject.html#pypdf._page.PageObject.extra
 You can use visitor-functions to control which part of a page you want to process and extract. The visitor-functions you provide will get called for each operator or for each text fragment.
 
 The function provided in argument visitor_text of function extract_text has five arguments:
-text, current transformation matrix, text matrix, font-dictionary and font-size.
-In most cases the x and y coordinates of the current position
-are in index 4 and 5 of the current transformation matrix.
+* text : the current text (as long as possible, can be up to a full line)
```
Comment (MartinThoma) [inline docs/user/extract-text.md:30]:
```suggestion
* text: the current text (as long as possible, can be up to a full line)
```

File: docs/user/extract-text.md [hunk 4]
```diff
@@ -27,14 +27,26 @@ Refer to [extract\_text](../modules/PageObject.html#pypdf._page.PageObject.extra
...
+* user_matrix: current matrix to move from user coordinate space(aka. CTM)
+* tm_matrix: current matrix from text coordinate space
+* font-dictionary: full font dictionary
+* font-size: the size (in text coordinate space)
+
+the matrix stores 6 parameters. the 4 first provides the rotation/scaling matrix and the last two provides the translation (horizontal/vertical)
```
Comment (MartinThoma) [inline docs/user/extract-text.md:36]:
```suggestion
The matrix stores 6 parameters. The first 4 provide the rotation/scaling matrix and the last two provide the translation (horizontal/vertical)
```

File: docs/user/extract-text.md [hunk 5]
```diff
@@ -27,14 +27,26 @@ Refer to [extract\_text](../modules/PageObject.html#pypdf._page.PageObject.extra
...
+it is recommended to use the user_matrix as it takes into all transformations.
```
Comment (MartinThoma) [inline docs/user/extract-text.md:37]:
```suggestion
It is recommended to use the user_matrix as it takes into all transformations.
```

File: docs/user/extract-text.md [hunk 6]
```diff
@@ -27,14 +27,26 @@ Refer to [extract\_text](../modules/PageObject.html#pypdf._page.PageObject.extra
...
+
+notes :
+ - as indicated in pdf 1.7 refeence, page 204 the user matrix applies to text space/image space/form space/pattern space.
+ - if you want to get the full transformation from text to user space, you can use the mult function (availalbe in global import) as follow :
+`txt2user = mult(tm, cm))`
+the font-size is the raw text size, that is affected by the user_matrix
```
Comment (MartinThoma) [inline docs/user/extract-text.md:43]:
```suggestion
Notes :

 - as indicated in the PDF 1.7 reference, page 204 the user matrix applies to text space/image space/form space/pattern space.
 - if you want to get the full transformation from text to user space, you can use the `mult` function (availalbe in global import) as follows:
`txt2user = mult(tm, cm))`
The font-size is the raw text size, that is affected by the `user_matrix`
```

**Comment by MartinThoma** [comment]
The cm/tm parts of the visitor functions are really hard to get right. Thank you for taking care of it @pubpub-zz :pray:

**Pull request merged**